cpp/kinematic_icp/local_map/SparseVoxelGrid.cpp

```diff
@@ -0,0 +1,148 @@
+    Bonxai::CoordT{0, 1, -1},   Bonxai::CoordT{0, 1, 0},   Bonxai::CoordT{0, 1, 1},
+
+    Bonxai::CoordT{1, -1, -1},  Bonxai::CoordT{1, -1, 0},  Bonxai::CoordT{1, -1, 1},
+    Bonxai::CoordT{1, 0, -1},   Bonxai::CoordT{1, 0, 0},   Bonxai::CoordT{1, 0, 1},
+    Bonxai::CoordT{1, 1, -1},   Bonxai::CoordT{1, 1, 0},   Bonxai::CoordT{1, 1, 1}};
+
+static constexpr uint8_t inner_grid_log2_dim = 2;
+static constexpr uint8_t leaf_grid_log2_dim = 3;
+}  // namespace
+
+namespace kinematic_icp {
+
+SparseVoxelGrid::SparseVoxelGrid(const double voxel_size,
+                                 const double clipping_distance,
+                                 const unsigned int max_points_per_voxel)
+    : voxel_size_(voxel_size),
+      clipping_distance_(clipping_distance),
+      max_points_per_voxel_(max_points_per_voxel),
+      map_(voxel_size, inner_grid_log2_dim, leaf_grid_log2_dim),
+      accessor_(map_.createAccessor()) {}
+
+std::tuple<Eigen::Vector3d, double> SparseVoxelGrid::GetClosestNeighbor(
+    const Eigen::Vector3d &query) const {
+    Eigen::Vector3d closest_neighbor = Eigen::Vector3d::Zero();
+    double closest_distance = std::numeric_limits<double>::max();
+    const auto const_accessor = map_.createConstAccessor();
+    const Bonxai::CoordT voxel = map_.posToCoord(query);
+    std::for_each(shifts.cbegin(), shifts.cend(), [&](const Bonxai::CoordT &voxel_shift) {
+        const Bonxai::CoordT query_voxel = voxel + voxel_shift;
+        const VoxelBlock *voxel_points = const_accessor.value(query_voxel);
+        if (voxel_points != nullptr) {
+            const Eigen::Vector3d &neighbor =
+                *std::min_element(voxel_points->cbegin(), voxel_points->cend(),
+                                  [&](const auto &lhs, const auto &rhs) {
+                                      return (lhs - query).norm() < (rhs - query).norm();
+                                  });
+            double distance = (neighbor - query).norm();
+            if (distance < closest_distance) {
+                closest_neighbor = neighbor;
+                closest_distance = distance;
+            }
+        }
+    });
+    return std::make_tuple(closest_neighbor, closest_distance);
+}
+
+void SparseVoxelGrid::AddPoints(const std::vector<Eigen::Vector3d> &points) {
+    const double map_resolution = std::sqrt(voxel_size_ * voxel_size_ / max_points_per_voxel_);
+    std::for_each(points.cbegin(), points.cend(), [&](const Eigen::Vector3d &p) {
+        const auto voxel_coordinates = map_.posToCoord(p);
+        VoxelBlock *voxel_points = accessor_.value(voxel_coordinates, /*create_if_missing=*/true);
+        if (voxel_points->size() == max_points_per_voxel_ ||
+            std::any_of(voxel_points->cbegin(), voxel_points->cend(), [&](const auto &voxel_point) {
+                return (voxel_point - p).norm() < map_resolution;
+            })) {
+            return;
+        }
+        voxel_points->reserve(max_points_per_voxel_);
```

Comment:
There will be duplicated calls to `reserve`, but since `max_points_per_voxel_` is fixed, this should not matter, right?